styleguide_example/blog_examples/google_login_server_flow/raw/service.py

```diff
@@ -0,0 +1,125 @@
+from random import SystemRandom
+from typing import Any, Dict, Tuple
+from urllib.parse import urlencode
+
+import jwt
+import requests
+from attr import define
+from django.conf import settings
+from django.core.exceptions import ImproperlyConfigured, ValidationError
+from django.urls import reverse_lazy
+from oauthlib.common import UNICODE_ASCII_CHARACTER_SET
+
+
+@define
+class GoogleRawLoginCredentials:
+    client_id: str
+    client_secret: str
+
+
+class GoogleRawLoginFlowService:
+    API_URI = reverse_lazy("api:google-oauth2:login-raw:callback-raw")
+
+    GOOGLE_AUTH_URL = "https://accounts.google.com/o/oauth2/v2/auth"
+    SCOPES = [
+        "https://www.googleapis.com/auth/userinfo.email",
+        "https://www.googleapis.com/auth/userinfo.profile",
+        "openid",
+    ]
+
+    GOOGLE_ID_TOKEN_INFO_URL = "https://www.googleapis.com/oauth2/v3/tokeninfo"
+    GOOGLE_ACCESS_TOKEN_OBTAIN_URL = "https://oauth2.googleapis.com/token"
+    GOOGLE_USER_INFO_URL = "https://www.googleapis.com/oauth2/v3/userinfo"
+
+    def __init__(self):
+        self._credentials = google_raw_login_get_credentials()
+
+    @staticmethod
+    def _jwt_decode_token(token: str) -> Dict[str, Any]:
+        decoded_data = jwt.decode(jwt=token, options={"verify_signature": False})
+        return decoded_data
+
+    @staticmethod
+    def _generate_state_session_token(length=30, chars=UNICODE_ASCII_CHARACTER_SET):
+        # This is how it's implemented in the official SDK
+        rand = SystemRandom()
+        state = "".join(rand.choice(chars) for _ in range(length))
+        return state
+
+    def _get_redirect_uri(self):
+        domain = settings.BASE_BACKEND_URL
+        api_uri = self.API_URI
+        redirect_uri = f"{domain}{api_uri}"
+        return redirect_uri
+
+    def get_authorization_url(self):
+        redirect_uri = self._get_redirect_uri()
+
+        state = self._generate_state_session_token()
+
+        params = {
+            "response_type": "code",
+            "client_id": self._credentials.client_id,
+            "redirect_uri": redirect_uri,
+            "scope": " ".join(self.SCOPES),
+            "state": state,
+            "access_type": "offline",
+            "include_granted_scopes": "true",
+            "prompt": "select_account",
+        }
+
+        query_params = urlencode(params)
+        authorization_url = f"{self.GOOGLE_AUTH_URL}?{query_params}"
+
+        return authorization_url, state
+
+    def get_access_credentials(self, *, code: str) -> Tuple:
```

Comment:
Lets name this simply - `get_tokens`.

The return type should be `Tuple[str, str]`, or we can even go one step further and create a type:

```python

@define
class GoogleAccessTokens:
    id_token: str
    access_token: str
```

And then we can actually put the method for decoding there :thinking: